Source/Core/Core/HW/EXI_DeviceAGP.cpp

```diff
@@ -0,0 +1,253 @@
+// Copyright 2013 Dolphin Emulator Project
+// Licensed under GPLv2
+// Refer to the license.txt file included.
+
+#include "Common/FileUtil.h"
+#include "Common/MemoryUtil.h"
+#include "Common/StdMakeUnique.h"
+#include "Core/ConfigManager.h"
+#include "Core/Core.h"
+#include "Core/HW/EXI_Device.h"
+#include "Core/HW/EXI_DeviceAGP.h"
+#include "Core/HW/Memmap.h"
+
+CEXIAgp::CEXIAgp(int index) :
+	m_slot(index)
+{
+	// Create the ROM
+	m_pHashArray = (u8*)AllocateMemoryPages(HASH_SIZE);
+	m_rom_size = 0;
+
+	LoadRom();
+
+	m_address = 0;
+	m_rom_hash_loaded = false;
+}
+
+CEXIAgp::~CEXIAgp()
+{
+	m_pROM = nullptr;
+	m_pHashArray = nullptr;
+}
+
+void CEXIAgp::DoHash(u8* data, u32 size)
+{
+	for (u32 it = 0; it < size; it++)
+	{
+		m_hash = m_hash ^ data[it];
+		m_hash = m_pHashArray[m_hash];
+	}
+}
+
+void CEXIAgp::LoadRom()
+{
+	// Load whole ROM dump
+	std::string path;
+	std::string filename;
+	std::string ext;
+	std::string gbapath;
+	SplitPath(m_slot == 0 ? SConfig::GetInstance().m_strGbaCartA : SConfig::GetInstance().m_strGbaCartB, &path, &filename, &ext);
+	gbapath = path + filename;
+	LoadFileToROM(gbapath + ext);
+	INFO_LOG(EXPANSIONINTERFACE, "Loaded gba rom: %s card: %d", gbapath.c_str(), m_slot);
+	LoadFileToEEPROM(gbapath + ".sav");
+	INFO_LOG(EXPANSIONINTERFACE, "Loaded gba sav: %s card: %d", gbapath.c_str(), m_slot);
+}
+
+void CEXIAgp::LoadFileToROM(std::string filename)
+{
+	File::IOFile pStream(filename, "rb");
+	if (pStream)
+	{
+		u64 filesize = pStream.GetSize();
+		m_rom_size = filesize & 0xFFFFFFFF;
+		m_rom_mask = (m_rom_size - 1);
+
+		m_pROM = (u8*)AllocateMemoryPages(m_rom_size);
```

Comment:
Make this a std::vector, and rename to m_rom. Then you can also likely get rid of m_rom_size (and compute m_rom_mask dynamically when you need it).